python/pyspark/errors/utils.py

```diff
@@ -119,3 +124,62 @@ def get_message_template(self, error_class: str) -> str:
             message_template = main_message_template + " " + sub_message_template
 
         return message_template
+
+
+def _capture_call_site(func_name: str) -> None:
+    """
+    Capture the call site information including file name, line number, and function name.
+
+    This function updates the thread-local storage from server side (PySparkCurrentOrigin)
+    with the current call site information when a PySpark API function is called.
+
+    Parameters
+    ----------
+    func_name : str
+        The name of the PySpark API function being captured.
+
+    Notes
+    -----
+    The call site information is used to enhance error messages with the exact location
+    in the user code that led to the error.
+    """
+    from pyspark.sql.session import SparkSession
+
+    spark = SparkSession._getActiveSessionOrCreate()
+    assert spark._jvm is not None
+
+    stack = inspect.stack()
+    frame_info = stack[-1]
+    function = func_name
+    filename = frame_info.filename
+    lineno = frame_info.lineno
+    call_site = f'"{function}" was called from\n{filename}:{lineno}'
```

Comment:
I'm not sure if we should use `fragment` from JVM instead of Python function name??

Reply:
Let me extract `fragment` from PySpark function name for consistency.